### ISSUE TITLE
[NEXT-1304] <Script> with `beforeInteractive` strategy ignores additional attributes in app router

### ISSUE THREAD

**Issue by victorandree** [vercel/next.js#49830]
_Note: Updated on August 24, 2023, with latest version of Next.js._

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000
    Binaries:
      Node: 20.5.1
      npm: 9.8.0
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.20-canary.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Script optimization (next/script)

### Link to the code that reproduces this issue

https://github.com/victorandree/next-script-before-interactive-attributes-app-dir

### To Reproduce

Setup a project with the app directory and in the root layout, use `<Script>` with `strategy="beforeInteractive"` and pass additional attributes. The additional attributes are not included in the resulting script tag.

### Describe the Bug

When using the `<Script>` tag from `next/script` in the root layout of the app directory, additional attributes are not included if you use `strategy="beforeInteractive"`:

```jsx
import Script from 'next/script';

export default function RootLayout({ children }) {
  return (
    <html>
      <head />
      <body>
        {children}
        <Script
          strategy="beforeInteractive"
          src="https://example.com/script.js"
          id="example-script"
          nonce="XUENAJFW"
          data-test="script"
        />
      </body>
    </html>
  );
}
```

A second point is that the `<script>` tag is not part of the HTML from the server, but is only added as (1) a `<link rel="preload">` tag and (2) to a `self.__next_s` array and (3) as part of server component 'chunks', before it's rendered into the DOM. 

Including such a script tag in `pages/_document` will render it with additional attributes (not reproduced) and as a script tag "directly". This issue therefore only affects the app directory.

A possible workaround is to include the `script` tags directly in `head` of your root layout:

```jsx
export default function RootLayout({ children }) {
  return (
    <html>
      <head>
        <script
          src="https://example.com/script.js"
          id="example-script"
          nonce="XUENAJFW"
          data-test="script"
        />
      </head>
      <body>{children}</body>
    </html>
  );
}
```

For visibility, I encountered this while trying to install the OneTrust cookie consent snippet on a Next.js site, which requires passing an identifier as `data-domain-script` to a `script` tag.

### Expected Behavior

The documented behavior of the `<Script>` tag is that it includes additional attributes (see ["Additional Attributes" on the "Script Optimization" page of the App router documentation](https://nextjs.org/docs/app/building-your-application/optimizing/scripts#additional-attributes)).

I also think it's a bit strange that the script tag is not rendered directly to the HTML from the server, since "before interactive" scripts are typically used for consent management (that's the example use case provided by the documentation), where it's important that they're inserted early.

I'd expect additional attributes to be included, even if I use it in the app directory.

<sub>[NEXT-1304](https://linear.app/vercel/issue/NEXT-1304/script-with-beforeinteractive-strategy-ignores-additional-attributes)</sub>

**Comment by maxstoyanov** [comment]
I just ran into this issue with 13.4.4.

> I also think it's a bit strange that the script tag is not rendered directly to the HTML from the server, since "before interactive" scripts are typically used for consent management (that's the example use case provided by the documentation), where it's important that they're inserted early.

I was quite surprised about this. The `beforeInteractive` strategy should include the script early in the head tag. It must be written without running scripts in the main thread.

**Comment by prepwork** [comment]
Next 13.4.4

Debugging half a day, I have several points to add. All with App Directory:

1. Script with `beforeInteractive` is successfully added inside `<head>` as described in docs, but **only** in DEV mode. In PROD (`next start`) it just sits in form of `self.__next_s=self.__next_s||[]).push...` as mentioned by @victorandree.

> A possible workaround is to include the script tags directly in head of your root layout

2.  This also works in development, but not in production. Even though <script> is being added in DOM by next in the manner smilar to `beforeInteractive` (e.g. `self.__next_s=self.__next_s||[]).push...`) it is not executed.

**Comment by victorandree** [comment]
@prepwork Hm, I just tried this out on 13.4.5, including the workaround (see https://github.com/victorandree/next-script-before-interactive-attributes-app-dir/commit/3140d3975fd74cf9a5905897ecf348ef702d35a6).

1. The bug described in the issue is still present, i.e. additional attributes are not passed to the (eventual) script tag. Unlike your first point, I am seeing a `script` tag in `head` once Next.js has processed `self.__next_s`. 
2. To me, the workaround seems to work even in production (i.e. `npm run build && npm run start`).

For reference, this is what the start of my server-rendered HTML looks like (when running HEAD of my reproduction repository using `npm run start`). I've formatted the HTML for readability:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <link rel="preload" as="script" href="https://example2.com/script2.js" />
    <link rel="preload" as="script" href="https://example.com/script.js" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <script src="https://example2.com/script2.js" data-test="script2"></script>
    <script
      src="/_next/static/chunks/polyfills-78c92fac7aa8fdd8.js"
      nomodule=""
    ></script>
  </head>
  <body>
    <script>
      (self.__next_s = self.__next_s || []).push([
        'https://example.com/script.js',
      ]);
    </script>
```

**Comment by victorandree** [comment]
@prepwork I tried this out, and if I add `useSearchParams`, I get this warning:

> Entire page / deopted into client-side rendering. https://nextjs.org/docs/messages/deopted-into-client-rendering /

I can verify that when this happens, the `self.__next_s` script is never "converted" into a script tag. The workaround _is_ converted (with additional attributes), but is not part of the HTML sent from the server.

IMO, this is technically another bug - fundamentally the same issue as #48335 (which is about static rendering of meta tags). The workaround there is to wrap `useSearchParams` in `Suspense` (which I don't think is obvious).

**Comment by prepwork** [comment]
@victorandree, thanks a lot! It helped me greatly!

**Comment by DanielSpindler** [comment]
@prepwork i dont know if this has to do with a recent change, but you should also consider this from nextJS https://nextjs.org/docs/messages/no-before-interactive-script-outside-document Seems to be  not really a bug, its just not intended that way

**Comment by desnor** [comment]
> @prepwork i dont know if this has to do with a recent change, but you should also consider this from nextJS https://nextjs.org/docs/messages/no-before-interactive-script-outside-document Seems to be not really a bug, its just not intended that way

with regard to this, is this still relevant when using the app router? it seems to me like something that just hasn't been updated yet as it seems conflicting to try and include a `pages/_document.tsx` to define the base when also using a `layout.tsx` root layout file. Does anyone know what the story is with this?

**Comment by victorandree** [comment]
@desnor @prepwork I agree that this isn't relevant for the app router. I don't see that warning when using the app router. The app router docs for `<Script>` even contain an example of using `beforeInteractive` in the root layout, so I definitely think this is supported: https://nextjs.org/docs/app/api-reference/components/script#beforeinteractive.

FWIW, I've updated my reproduction repository to next@13.4.20-canary.4 and the issue is still present.

**Comment by azar-xp** [comment]
This seems to be a bug.

Official docs for /app [https://nextjs.org/docs/app/api-reference/components/script#beforeinteractive](https://nextjs.org/docs/app/api-reference/components/script#beforeinteractive) says that it's before and after interactive is supported from next 13.0.0 but not for build script.

Proposed fix on docs for /app [https://nextjs.org/docs/messages/no-before-interactive-script-outside-document](https://nextjs.org/docs/messages/no-before-interactive-script-outside-document) is not relevant because it's conflicting with /app/layout.tsx and not working.

Workarounds do not work as well. This needs to be fixed from next side.

**Comment by djacquemard** [comment]
Hello ! 

Same issue here trying to add OneTrust script with beforeInteractive strategy.
Thank you for the workaround @victorandree !

Did you add the OneTrust Cookie List to your website ? On my side, trying to add  `<div id="ot-sdk-cookie-policy" />` container cause (of course) hydration error and I can't solve this by any way.

**Comment by shaunforage** [comment]
Experiencing the same issue on 13.5.6.

**Comment by jesse-mm** [comment]
Confirmed still not working as advertised in `14.0.2`. I've put the script I want to load in the `RootLayout` as described in the docs. But it still loads next modules first, after that it injects the custom script.

Following the documentation this should not happen:
> `beforeInteractive`: Load before any Next.js code and before any page hydration occurs.

**Comment by fmontes** [comment]
Can confirm this is happening in 14 for me as well.

**Comment by zbauman3** [comment]
I'm on `14.0.4` and can confirm this is still an issue.

**Comment by zbauman3** [comment]
I've created a PR to fix this issue. I've tested locally and can confirm this solves the problem. Now just waiting on Next.js folks to review.

https://github.com/vercel/next.js/pull/59779
